src/Filesystem/Dir.php

```
@@ -588,7 +588,11 @@ public static function remove(string $dir): bool
 			}
 		}
 
-		return rmdir($dir);
+		return Helpers::handleErrors(
+			fn (): bool => rmdir($dir),
+			fn (int $errno, string $errstr) => true,
+			fn (): never => throw new Exception('The directory could not be deleted'),
+		);
```

I read a few posts on how to deal with such race conditions. Some recommendations were:
- Rename the directory first as this is atomic. Then delete the renamed folder. This way any new writes to the same location of the folder that come in after `Dir::remove()` started can still be created, but will not create problems deleting the files captured at the start of the remove call.
- Instead of just giving up and throwing an exception, there are a few posts that recommend to keep retrying the `rmdir` call a few times if an error occurs. That increases the chances of one being completed. However, in a high traffic caching environment, there is still the chance that none will actually succeed.

What do you think @lukasbestle  @bastianallgeier?

---

I like the idea to think outside the box for more ways to prevent this. Retrying seems worth a shot. Moving the folder sounds like a good idea at first too, but we could end up with a lot of unremoved clones if we don't get it right.

---

I really like the idea. The `rename` method will return either true or false. If it returns true, I see no reason why `Dir::remove` shouldn't work. If `rename` returns false, the `Dir::remove` process won't even start. Thus, no clone will be created.

If we want to be absolutely sure, we can add an extra clone check and delete handler inside the `FileCache::flush()` method. This way, if a cloned folder remains somehow, it will always be cleaned up before flushing.

---

Making the deletion atomic via a prior renaming operation would be perfect indeed. But it needs to be reliable (no left-over random directories on failure) and it also needs to be transparent to calling code (there should be no need to do anything special in `FileCache::flush()` for example).

With a naive approach of just renaming and then deleting contents as normal, we could run into issues e.g. when any file is not writable by the PHP user. The deletion will then fail in the middle of the operation and the renamed directory will still be there.

What we would need is logic along these lines:

```php
rename($currentName, $newName = $currentName . $randomSuffix);

try {
  // current `scandir()` loop
} catch (Throwable $e) {
  Helpers::handleErrors(
    fn (): bool => rename($newName, $currentName),
    function () {
      // someone has already created the original directory again;
      // instead move all contents of the renamed directory to the new directory
    }
  );
}
```

But that's very complex and would require us to convert all PHP warnings that can occur in the process to proper exceptions. Which would be nice anyway, but a project on its own.

---

Another thought: `FileCache::flush()` doesn't really need to delete the directories (neither the top-level one nor any subfolders). These can still be deleted on the next flush operation without any impact to the cache behavior. So we could add a `bool $lenient = false` mode to `Dir::remove()` that uses `@rmdir()` instead of `rmdir()`. This way, any issues with file deletions will still trigger a warning, but directory removals will no longer make the process fail.

---

Just another thought on the renaming: could we rename it to somewhere where it doesn't harm if by whatever reason it remains as leftover? And then e.g. via our Panel garbage collection just regularly delete any potential leftovers in that location? (As anything in that location would have been meant to be deleted anyways)

---

If the tmp folder would be cleaned up correctly by the system, it would be ideal to move it there if possible. But we have a similar situation with uploads already. It's not being cleaned up as regular as we wish and as long as we don't have some built in queue system, I'm not sure what would trigger such clean ups on our end.

---

Panel garbage collection which runs in 10% of any Panel requests. The same we use for uploads.

---

Ah, that's a good idea indeed. Maybe we should separate this. I think @afbora's PR is already a great first step to improve error handling. Adding a second layer to solve the flush problem would still make sense to me, but could be another PR. What do you think?

---

Can be two separate PRs, yes. But wouldn't use the rename or other solutions just for flushing but `Dir::remove()` in general as well.

---

I agree on the two-step approach.

However I think this PR can be considered a breaking change. Before, any failure during the flushing operation will trigger a warning and therefore an exception if Whoops is enabled. Now, there will be only a silent `false` return value that so far no one was checking. So if the caller needs to rely on the cache being flushed for following actions, such use cases will now silently break if anything goes wrong during the flushing operation.

IMO to fix this in a way that still keeps reliability, we need to differentiate between errors during file deletion and during directory deletion (the latter can be ignored).

---

I think we need to view this from the opposite angle to be honest. Right now, you can no longer save pages or anything else on a high-traffic website when the folder cannot be properly removed, because it will always cause an error. This is pretty bad if you ask me. I still agree that we cannot simply swallow the issue. We should maybe think about writing it to the php error log instead though.

---

I think both of you are right: We need to fix this, but maybe the fix needs to be more specific about which errors get swallows (and maybe written to the php error log) and which ones it should not swallow.

Cause this isn't just about flushing the chance, right. Any change will affect also all other use cases of `Dir::remove()`.

---

That's true. Let's maybe break it down. 

I think it's a good idea to improve the error handling in Dir::remove. This will be helpful for all situations in which it is being used. 

We are not swallowing anything here, we only make it easier to catch because it will result in an exception and not a warning. 

Swallowing it in the flush method could indeed be the wrong approach at this point. Maybe we shouldn't be doing this in this PR at all and only focus on the error handling in the Dir method and go from here in the next step?